### PR TITLE
Move workspace listing as a separate task for an assessment workflow

### PR DESCRIPTION
**Pull request by HariGS-DB**

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
let's keep changes to src/databricks/labs/ucx/workspace_access/listing.py out of scope from this PR and only separate on the level of runtime.py

**Comment by HariGS-DB** [comment]
Does this looks better. I also noticed there is no test coverage for runtime.py so two checks are failing
@nfx

File: src/databricks/labs/ucx/runtime.py [hunk 1]
```diff
@@ -203,6 +204,18 @@ def assessment_report(_: WorkspaceConfig):
...
+@task("assessment", depends_on=[setup_schema])
+def workspace_listing(cfg: WorkspaceConfig, start_path: str = "/"):
+    """Scans the workspace for workspace objects. It recursively list all sub directories
+    and compiles a list of directories, notebooks, files, repos and libraries in the workspace.
+
+    Its uses multi-threading to parallelize the listing process to speed up on big workspaces.
+    It accepts starting path as the parameter defaulted to the root path '/'."""
+    ws = WorkspaceClient(config=cfg.to_databricks_config())
+    ws_listing = WorkspaceListing(ws, num_threads=cfg.num_threads)
+    ws_listing.walk(start_path)
```
Comment (nfx) [inline src/databricks/labs/ucx/runtime.py:216]: and how would `crawl_permissions` know about the results of it? make it extend the `CrawlerBase` :)

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/runtime.py (hunk 1 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/runtime.py]: have done the changes. please check. I have not touched the 
https://github.com/databrickslabs/ucx/blob/main/src/databricks/labs/ucx/workspace_access/generic.py
WorkspaceListing class which is doing the workspace listing. Should we change that to fetch from the new crawler?
@nfx

File: pyproject.toml [hunk 2]
```diff
@@ -35,6 +35,7 @@ dependencies = [
 test = [
     "coverage[toml]>=6.5",
     "pytest",
+    "pyspark",
```
Comment (nfx) [inline pyproject.toml:38]: you are not allowed to add `pyspark` as dependency. use `sql_backend` fixtures.

File: src/databricks/labs/ucx/runtime.py [hunk 3]
```diff
@@ -203,6 +204,18 @@ def assessment_report(_: WorkspaceConfig):
     dashboard _before_ all tasks have been completed, but then only already completed information is shown."""
 
 
+@task("assessment", depends_on=[setup_schema])
```
Comment (nfx) [inline src/databricks/labs/ucx/runtime.py:207]: and make refresh dashboard depend on it

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/runtime.py (hunk 3 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/runtime.py]: done

File: pyproject.toml (hunk 2 above)
Comment (HariGS-DB) [reply pyproject.toml]: removed dependency @nfx

**Review by nfx — CHANGES_REQUESTED** [review]
can you write some integration tests to make sure `PermissionManager` can pick it up and call the `snapshot()` method of it?

**Comment by HariGS-DB** [comment]
> can you write some integration tests to make sure `PermissionManager` can pick it up and call the `snapshot()` method of it?

Need confirmation. I need to update PermissionManager to use the new WorkspaceObjectCrawler and then write integration tests for it?

**Comment by nfx** [comment]
>Need confirmation. I need to update PermissionManager to use the new WorkspaceObjectCrawler and then write integration tests for it?

that's correct

**Comment by HariGS-DB** [comment]
> > Need confirmation. I need to update PermissionManager to use the new WorkspaceObjectCrawler and then write integration tests for it?
> 
> that's correct

should i create a WorkspaceObjectSupport class like for scim, redash, tacl or use the existing GenericSupport and update the WorkspaceListing to use the new crawler?

**Comment by nfx** [comment]
@HariGS-DB also please add a documentation about this table to `docs/table_persistence.md`

**Comment by HariGS-DB** [comment]
@nfx added integration test and tested from integration/assessment/test_assessment.py and  integration/workspace_access/test_workspace_access.py
Also added documentation for the new table

File: src/databricks/labs/ucx/runtime.py [hunk 4]
```diff
@@ -165,7 +166,19 @@ def assess_global_init_scripts(cfg: WorkspaceConfig):
     crawler.snapshot()
 
 
-@task("assessment", depends_on=[crawl_grants])
+@task("assessment", depends_on=[setup_schema])
+def workspace_listing(cfg: WorkspaceConfig):
+    """Scans the workspace for workspace objects. It recursively list all sub directories
+    and compiles a list of directories, notebooks, files, repos and libraries in the workspace.
+
+    Its uses multi-threading to parallelize the listing process to speed up on big workspaces.
```
Comment (nfx) [inline src/databricks/labs/ucx/runtime.py:174]:
```suggestion
    It uses multi-threading to parallelize the listing process to speed up execution on big workspaces.
```

File: src/databricks/labs/ucx/runtime.py [hunk 5]
```diff
@@ -165,7 +166,19 @@ def assess_global_init_scripts(cfg: WorkspaceConfig):
...
+    It accepts starting path as the parameter defaulted to the root path '/'."""
+    ws = WorkspaceClient(config=cfg.to_databricks_config())
+    crawler = WorkspaceListing(ws, RuntimeBackend(), cfg.inventory_database)
```
Comment (nfx) [inline src/databricks/labs/ucx/runtime.py:177]: Missing workspace start path from configuration

File: src/databricks/labs/ucx/workspace_access/generic.py [hunk 6]
```diff
@@ -208,38 +217,72 @@ def _prepare_new_acl(
...
+            table="workspace_objects",
+            klass=WorkspaceObjectInfo,
+        )
         self._ws = ws
         self._num_threads = num_threads
         self._start_path = start_path
+        self._sql_backend = sql_backend
+        self._inventory_database = inventory_database
+
+    def _crawl(self) -> list[WorkspaceObjectInfo]:
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/generic.py:244]: Lines 245-248 should be removed, this is one method :)

File: src/databricks/labs/ucx/workspace_access/generic.py [hunk 7]
```diff
@@ -208,38 +217,72 @@ def _prepare_new_acl(
...
+
+    def _assess_workspace_listing(self):
+        from databricks.labs.ucx.workspace_access.listing import WorkspaceListing
+
+        ws_listing = WorkspaceListing(self._ws, num_threads=self._num_threads, with_directories=False)
+        for _object in ws_listing.walk(self._start_path):
+            workspace_object_info = WorkspaceObjectInfo(
+                _object.object_type.name, str(_object.object_id), _object.path, _object.language
+            )
+            yield workspace_object_info
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/generic.py:255]: Can you inline the variable? It's not used anywhere else

File: src/databricks/labs/ucx/workspace_access/manager.py [hunk 8]
```diff
@@ -46,7 +46,13 @@ def factory(
             generic.Listing(generic.experiments_listing(ws), "experiment_id", "experiments"),
             generic.Listing(generic.models_listing(ws), "id", "registered-models"),
             generic.Listing(generic.tokens_and_passwords, "object_id", "authorization"),
-            generic.WorkspaceListing(ws, num_threads=num_threads, start_path=workspace_start_path),
+            generic.WorkspaceListing(
+                ws,
+                sql_backend=sql_backend,
+                inventory_database=inventory_database,
+                num_threads=num_threads,
+                start_path=workspace_start_path,
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/manager.py:54]: Start path and num threads only have the effect in runtime.py, add them there and not here

File: tests/integration/assessment/test_assessment.py [hunk 9]
```diff
@@ -165,3 +166,18 @@ def test_spn_crawler_with_available_secrets(
     ws.secrets.delete_secret(scope=secret_scope, key=secret_key)
 
     assert len(results) >= 2
+
+
+def test_workspace_object_crawler(ws, make_notebook, make_directory, inventory_schema, sql_backend):
+    new_notebook = make_notebook()
+    new_directory = make_directory()
+    workspace_listing = WorkspaceListing(ws=ws, sql_backend=sql_backend, inventory_database=inventory_schema)
```
Comment (nfx) [inline tests/integration/assessment/test_assessment.py:174]: Add start path, otherwise this test will scan the entire workspace

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/workspace_access/manager.py (hunk 8 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/workspace_access/manager.py:54]: This is also invoked by GroupMigrationToolkit, shouldn't the num_thread and start_path used by that call to set?

File: src/databricks/labs/ucx/workspace_access/generic.py (hunk 6 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/workspace_access/generic.py:244]: added

File: src/databricks/labs/ucx/workspace_access/generic.py (hunk 7 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/workspace_access/generic.py]: done

File: tests/integration/assessment/test_assessment.py (hunk 9 above)
Comment (HariGS-DB) [reply tests/integration/assessment/test_assessment.py]: done

File: src/databricks/labs/ucx/workspace_access/generic.py [hunk 10]
```diff
@@ -208,38 +217,66 @@ def _prepare_new_acl(
...
+            backend=sql_backend,
+            catalog="hive_metastore",
+            schema=inventory_database,
+            table="workspace_objects",
+            klass=WorkspaceObjectInfo,
+        )
         self._ws = ws
         self._num_threads = num_threads
         self._start_path = start_path
+        self._sql_backend = sql_backend
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/generic.py:241]: You don't have to assign these two fields, it's already done by crawlerbase

**Review by nfx — APPROVED** [review]

**Review by nfx — APPROVED** [review]

File: src/databricks/labs/ucx/workspace_access/generic.py (hunk 10 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/workspace_access/generic.py:241]: good point

Comment (HariGS-DB) [reply src/databricks/labs/ucx/workspace_access/generic.py:241]: will update in the next issue PR

**Pull request merged**